### DISCUSSION TITLE
How to extract ordered nodes of a connected path

### DISCUSSION THREAD

**Discussion by griff10000** [networkx/networkx#4533] [General]
After creating a connected graph from unordered edges, the code:
```
import networkx as nx

G = nx.Graph()
G.add_edges_from(edges)
nx.draw(G)
```
creates a nice graph of ordered nodes.
Q: how do I obtain a list of the ordered nodes used to draw the graph?

**Comment by dschult** [comment]
`G.nodes` provides an ordered read-only dict interface keyed by node to the node attributes. So, `list(G.nodes)` will give you a list of the nodes and `for node in G.nodes:` will iterate through the nodes as a loop.
The [main documentation tutorial](https://networkx.org/documentation/stable/tutorial.html) gives more information.

**Comment by griff10000** [comment]
Thanks for your reply.
My question should have been more clear. I am actually after the nodes arranged in the sequence shown by the `nx.draw()` command, e.g. the code:
```
import matplotlib.pyplot as plt
import networkx as nx

edges = [ [ 0,  1],
          [ 0,  3],
          [ 1,  2],
          [ 2,  6],
          [ 3,  7],
          [ 6, 11],
          [ 7, 12],
          [11, 15],
          [12, 16],
          [15, 18],
          [16, 17],
          [17, 18] ]


G = nx.Graph()
G.add_edges_from(edges)
nx.draw(G,with_labels=True)
plt.show()
```

Produces the attached diagram of connected nodes.
I am after the nodes as ordered in the diagram, i.e. the node sequence:

`[0, 1, 2, 6, 11, 15, 18, 17, 16, 12 ,7, 3]`

The command `list(G.nodes)` gives:
`[0, 1, 3, 2, 6, 7, 11, 12, 15, 16, 18, 17]`

Which is only sorted into numerical order.

Apologies if I am missing something simple.

![ex_plot](https://user-images.githubusercontent.com/49166905/104134950-cc1a5b80-5384-11eb-8dbe-19d4a6611723.png)

**Reply by dschult** [reply]
If I understand you correctly you are looking for the path around the circle.
But if this wasn't a path, there wouldn't necessarily be a path. I guess a general approach could be to find all_simple_paths between two nodes, turn them into lists and add the lists. But maybe a general approach is not what you need...

A simpler approach for just getting the nodes of a path from the edges of a path is to process your `edges` input:
```python
path = [edges[0][0]] + [e[1] for e in edges]
```
This is only going to give a path if the edges form a path.

**Comment by griff10000** [comment]
In my application I form a network over a spatial domain and I identify the boundary nodes. By definition a path exists, and I am seeking a simple way to identify this path. I was hoping that `networkx` would provide the solution.
The fact that the command `nx.draw()` is able to draw this path indicates that it must be calculated somewhere. Otherwise the graph would look like a random set of connected lines.
Is `nx.draw()` part of `networkx' or does it rely on another package?
Thank you for your replies.

**Comment by dschult** [comment]
`nx.draw` simply draws the edges and positions the nodes to spread out the network. It doesn't take into account paths while drawing. That said, networkx has a number of path finding algorithms. If you can identify the boundary nodes, say, u and v, then `nx.shortest_path(G, u, v)` will return the path.  In your example above:

```python
G = nx.Graph()
G.add_edges_from(edges)
u=edges[0][0]
v=edges[-1][-1]
path=nx.shortest_path(G, u, v)
```
No drawing is necessary.

**Comment by griff10000** [comment]
Thank you for your reply. Unfortunately this does not produce the complete path.
However, I hound a solution at:
[cycle_basis](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.cycles.cycle_basis.html#networkx.algorithms.cycles.cycle_basis)
The following code now produces the correct path which matches the above diagram:
```
import matplotlib.pyplot as plt
import networkx as nx

edges = [ [ 0,  1],
          [ 0,  3],
          [ 1,  2],
          [ 2,  6],
          [ 3,  7],
          [ 6, 11],
          [ 7, 12],
          [11, 15],
          [12, 16],
          [15, 18],
          [16, 17],
          [17, 18] ]


G = nx.Graph()
G.add_edges_from(edges)
nx.draw(G,with_labels=True)
plt.show()

Gcb = nx.cycle_basis(G,root=3)
print(Gcb)

```
i.e. the output is:
`[[0, 1, 2, 6, 11, 15, 18, 17, 16, 12, 7, 3]]`

Thanks again.
